JavaScript 2021 chapter and query updates (#2567)
HTTPArchive/almanac.httparchive.org: src/content/en/2021/javascript.md

---
#See https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide#metadata-to-add-at-the-top-of-your-chapters
title: JavaScript
description: JavaScript chapter of the 2020 Web Almanac covering the usage of JavaScript on the web, libraries and frameworks, compression, web components, and source maps.
authors: [NishuGoel]
reviewers: [soulcorrosion, mgechev, rviscomi, pankajparkar, tunetheweb]
analysts: [pankajparkar, max-ostapenko, rviscomi]
editors: [rviscomi, pankajparkar]
translators: []
results: https://docs.google.com/spreadsheets/d/1zU9rHpI3nC6jTz3xgN6w13afW7x34xAKBh2IPH-lVxk/
featured_quote: Features provided for improving the rendering time and resource loading time could be leveraged better to see the overall impact and get an even better experience with respect to the performance.
featured_stat_1: 4
featured_stat_label_1: Median asynchronous requests made per mobile page.
featured_stat_2: 45%
featured_stat_label_2: Images are the most requested content type on pages.
featured_stat_3: 2.7%

> featured_stat_label_3: Percent of desktop pages using custom elements.

Reviewer: ```suggestion
featured_stat_label_3: Percent of desktop pages using custom elements.
unedited: true
```

<figure markdown>
<table>
  <tr>
   <td>
   </td>
   <td><strong>Values</strong>
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><strong>client</strong>
   </td>
   <td><strong>async</strong>
   </td>
   <td><strong>defer</strong>
   </td>
   <td><strong>neither</strong>
   </td>
  </tr>
  <tr>
   <td>desktop
   </td>
   <td><p style="text-align: right">
89.26%</p>

   </td>
   <td><p style="text-align: right">
48.09%</p>

   </td>
   <td><p style="text-align: right">
10.27%</p>

   </td>
  </tr>
  <tr>
   <td>mobile
   </td>
   <td><p style="text-align: right">
89.12%</p>

   </td>
   <td><p style="text-align: right">
47.78%</p>

   </td>
   <td><p style="text-align: right">
10.40%</p>

   </td>
  </tr>
</table>

<figcaption>{{ figure_link(caption="Percent of pages using `async`, `defer`, or neither.", sheets_gid="2038121216", sql_file="breakdown_of_pages_using_async_defer.sql") }}</figcaption>
</figure>

There was an antipattern observed in the last year's results that some websites use both `async` and `defer` attribute resulting in fallback to `async` if the browser supports it.

> The usage was so frequent that [11.4%](https://almanac.httparchive.org/en/2020/javascript#how-do-we-load-our-javascript) of requests made on the mobile pages were seen with at least one script with `async` and `defer` attributes used together.  The [root causes](https://twitter.com/rick_viscomi/status/1331735748060524551?s=20) were found and an action item was also taken down to[ remove such usage going forward](https://twitter.com/Kraft/status/1336772912414601224?s=20). It’d be very interesting this year to look at where this stands.

Reviewer: ```suggestion
The usage was so frequent that [11.4%](https://almanac.httparchive.org/en/2020/javascript#how-do-we-load-our-javascript) of requests made on the mobile pages were seen with at least one script with `async` and `defer` attributes used together.  The [root causes](https://twitter.com/rick_viscomi/status/1331735748060524551?s=20) were found and an action item was also taken down to [remove such usage going forward](https://twitter.com/Kraft/status/1336772912414601224?s=20). It’d be very interesting this year to look at where this stands.
```

  )
}}

As it turns out, 2.6% of requests use this antipattern with both `async` and `defer` on the same script element.

### first party vs third party

The median number of JavaScript resources requested on desktop pages, as observed earlier, is 21. Let us now look into which ones of these are first party or third party requests. 

{{ figure_markup(
  image="js-requests-mobile-host.png",
  caption="Distribution of the number of JavaScript requests per mobile page by host",
  description="",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=395424439&format=interactive",
  sheets_gid="1354767842",
  sql_file="requests_by_3p.sql"
  )
}}

{{ figure_markup(
  image="js-requests-desktop-host.png",
  caption="Distribution of the number of JavaScript requests per desktop page by host.",
  description="",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1584404981&format=interactive",
  sheets_gid="1354767842",
  sql_file="requests_by_3p.sql"
  )
}}

> A median of desktop pages request 11 third-party resources whereas 10 first-party requests. This difference increases as we move up to the 90th percentile as 33 requests on mobile pages are first party but the number goes up to 34 for third-party requests for the mobile pages.  Clearly, the number of third party resources requested is always one step ahead of the first-party resources. Irrespective of the performance and reliability [risks that requesting third-party resources brings](https://css-tricks.com/potential-dangers-of-third-party-javascript/), the usage seems to favour third-party scripts, which could be due to the[ useful interactivity features](https://developers.google.com/web/fundamentals/performance/critical-rendering-path/adding-interactivity-with-javascript) that it gives to the web. 

Reviewer: ```suggestion
A median of desktop pages request 11 third-party resources whereas 10 first-party requests. This difference increases as we move up to the 90th percentile as 33 requests on mobile pages are first party but the number goes up to 34 for third-party requests for the mobile pages.  Clearly, the number of third party resources requested is always one step ahead of the first-party resources. Irrespective of the performance and reliability [risks that requesting third-party resources brings](https://css-tricks.com/potential-dangers-of-third-party-javascript/), the usage seems to favour third-party scripts, which could be due to the [useful interactivity features](https://developers.google.com/web/fundamentals/performance/critical-rendering-path/adding-interactivity-with-javascript) that it gives to the web. 
```

---
#See https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide#metadata-to-add-at-the-top-of-your-chapters
title: JavaScript

> description: JavaScript chapter of the 2020 Web Almanac covering the usage of JavaScript on the web, libraries and frameworks, compression, web components, and source maps.

Reviewer: ```suggestion
description: JavaScript chapter of the 2021 Web Almanac covering the usage of JavaScript on the web, libraries and frameworks, compression, web components, and source maps.
```

authors: [NishuGoel]
reviewers: [soulcorrosion, mgechev, rviscomi, pankajparkar, tunetheweb]
analysts: [pankajparkar, max-ostapenko, rviscomi]
editors: [rviscomi, pankajparkar]
translators: []

> results: https://docs.google.com/spreadsheets/d/1zU9rHpI3nC6jTz3xgN6w13afW7x34xAKBh2IPH-lVxk/

Reviewer: @NishuGoel could you add a bio for yourself? This is shown at the bottom of the chapter.
```suggestion
results: https://docs.google.com/spreadsheets/d/1zU9rHpI3nC6jTz3xgN6w13afW7x34xAKBh2IPH-lVxk/
NishuGoel_bio: TODO
```

NishuGoel_bio: Nishu Goel is an engineer at <a hreflang="en" href="http://webdataworks.io/">Web DataWorks</a>. She is a Google Developer Expert for Web Technologies and Angular, Microsoft MVP for Developer Technologies, and the author of Step by Step Guide Angular Routing (BPB, 2019) and A Hands-on Guide to Angular (Educative, 2021). Find her writings at <a hreflang="en" href="https://unravelweb.dev/">unravelweb.dev</a>.
featured_quote: Features provided for improving the rendering time and resource loading time could be leveraged better to see the overall impact and get an even better experience with respect to the performance.
featured_stat_1: 4
featured_stat_label_1: Median asynchronous requests made per mobile page.
featured_stat_2: 45%
featured_stat_label_2: Images are the most requested content type on pages.
featured_stat_3: 2.7%
featured_stat_label_3: Percent of desktop pages using custom elements.
unedited: true
---

### Introduction

> The speed and consistency that the JavaScript language has evolved with over the past years is tremendous. While in the past it was  used primarily for client-side, it has taken a very important and respected place in the world of building services and server-side tools. JavaScript has evolved to a point where it is not only possible to create faster applications but also to [run servers within browsers](https://blog.stackblitz.com/posts/introducing-webcontainers/).

Reviewer 2: ```suggestion
The speed and consistency that the JavaScript language has evolved with over the past years is tremendous. While in the past it was used primarily for client-side, it has taken a very important and respected place in the world of building services and server-side tools. JavaScript has evolved to a point where it is not only possible to create faster applications but also to [run servers within browsers](https://blog.stackblitz.com/posts/introducing-webcontainers/).
```

### How much JavaScript do we load?

To measure is the key towards improvement, they say. To improve the usage of JavaScript in our applications, we need to measure how much JavaScript being shipped is actually required. Let’s dig in to understand the distribution of JavaScript bytes per page, considering what a major role it plays in the web setup.

The 50th percentile (median) desktop page loads 463 kilobytes of JavaScript, whereas a median page load on a mobile device sends 427 kilobytes. 

{{ figure_markup(
  image="javascript-bytes-per-page.png",
  caption="Distribution of the amount of JavaScript kilobytes loaded per page.",
  description="Bar chart showing the distribution of JavaScript bytes per page. The median desktop page loads 463 kilobytes of JavaScript. The 10th, 25th, 50th, 75th, and 90th percentiles for desktop are: 94 KB, 220 KB, 463 KB, 852 KB, and 1,354 KB.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=329775434&format=interactive",
  sheets_gid="18398250",
  sql_file="bytes_2021.sql"
  )
}}

> Compared to [2019’s results](https://almanac.httparchive.org/en/2019/javascript#fig-2), this shows an increase of 18.4% in the usage of JavaScript for desktop devices and an increase of 18.9% on mobile devices. As we see the trend is towards using more JavaScript over the years, this could slow down the rendering of an application given the additional CPU work. It's worth to note that these statitics represent the  transferred bytes which could be compressed responses and thus, the actual cost to the CPU could be significantly higher. 

Reviewer 2: ```suggestion
Compared to [2019’s results](https://almanac.httparchive.org/en/2019/javascript#fig-2), this shows an increase of 18.4% in the usage of JavaScript for desktop devices and an increase of 18.9% on mobile devices. As we see the trend is towards using more JavaScript over the years, this could slow down the rendering of an application given the additional CPU work. It's worth to note that these statistics represent the transferred bytes which could be compressed responses and thus, the actual cost to the CPU could be significantly higher. 
```

### JS requests per page

One of the contributing factors towards slow rendering of the web page could be the requests made on the page, especially when they are blocking.  It would be interesting to look at the number of JavaScript requests made per page on both desktop and mobile devices.

{{ figure_markup(
  image="js-requests-per-page.png",
  caption="Distribution of JavaScript requests on desktop and mobile pages.",
  description="Bar chart showing the distribution of JavaScript requests on desktop and mobile pages. The median mobile page loads 20 JavaScript resources as compared to 21 JavaScript resources on desktop. The 10th, 25th, 50th, 75th, and 90th percentiles for requests made on mobile are: 4, 10, 20, 35, 56",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1487796178&format=interactive",
  sheets_gid="159538568",
  sql_file="requests_2021.sql"
  )
}}

> The median desktop page loads 21 JavaScript resources(.js and .mjs files), going up to 59 resources at the 90th percentile. 

Reviewer 2: ```suggestion
The median desktop page loads 21 JavaScript resources (`.js` and `.mjs` files), going up to 59 resources at the 90th percentile. 

```

Read more about the state of the protocols in the [HTTP](./http) chapter.

### module and nomodule

When loading a website, the browser renders the HTML and requests the appropriate resources. It consumes the polyfills referenced in the code for the effective rendering and functioning of the page. The modern browsers that support newer syntax like [arrow functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions), [async functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function), etc. do not need loads of polyfills to make things work and therefore, should not have to.

This is when differential loading takes care of things. Specifying the `type=”module”` attribute would serve the modern browsers the bundle with modern syntax and fewer polyfills, if any. Similarly, older browsers that lack support for modules will be served the bundle with required polyfills and transpiled code syntax with attribute `type=”nomodule”`. Read more about the usage of modue/nomodule [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#applying_the_module_to_your_html)

> Let's look at the data to understand the adoption of these attributes. \

Reviewer 2: ```suggestion
Let's look at the data to understand the adoption of these attributes.
```

4.6% of desktop pages use the attribute `type=”module”` whereas only 3.9% of mobile pages use `type=”nomodule”`. This could be due to the fact that the mobile dataset being much larger contains more "long-tail" websites that might not be using the latest features.

<figure markdown>
<table>
  <tr>
   <td><strong>client</strong>
   </td>
   <td><strong>module</strong>
   </td>
   <td><strong>nomodule</strong> 
   </td>
  </tr>
  <tr>
   <td><em>Desktop</em>
   </td>
   <td>4.6%
   </td>
   <td>3.9%
   </td>
  </tr>
  <tr>
   <td><em>Mobile</em>
   </td>
   <td>4.3%
   </td>
   <td>3.9%
   </td>
  </tr>
</table>

<figcaption>{{ figure_link(caption="Distribution of differential loading usage on desktop and mobile clients.", sheets_gid="1261294750", sql_file="module_and_nomodule.sql") }}</figcaption>
</figure>

> It is important to note that with the [end of support for IE11 browser](https://docs.microsoft.com/en-us/lifecycle/announcements/internet-explorer-11-support-end-dates), differential loading is now well supported across all browsers that most sites support.

Reviewer 2: ```suggestion
It is important to note that with the [end of support for IE11 browser](https://docs.microsoft.com/en-us/lifecycle/announcements/internet-explorer-11-support-end-dates), differential loading is less applicable because evergreen browsers support modern JavaScript syntax.
```

Reviewer 2: You can support this statement with https://github.com/angular/angular/issues/41840.

<figure markdown>
<table>
  <tr>
   <td>
   </td>
   <td><strong>Values</strong>
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><strong>client</strong>
   </td>
   <td><strong>async</strong>
   </td>
   <td><strong>defer</strong>
   </td>
   <td><strong>neither</strong>
   </td>
  </tr>
  <tr>
   <td>desktop
   </td>
   <td><p style="text-align: right">
89.26%</p>

   </td>
   <td><p style="text-align: right">
48.09%</p>

   </td>
   <td><p style="text-align: right">
10.27%</p>

   </td>
  </tr>
  <tr>
   <td>mobile
   </td>
   <td><p style="text-align: right">
89.12%</p>

   </td>
   <td><p style="text-align: right">
47.78%</p>

   </td>
   <td><p style="text-align: right">
10.40%</p>

   </td>
  </tr>
</table>

<figcaption>{{ figure_link(caption="Percent of pages using `async`, `defer`, or neither.", sheets_gid="2038121216", sql_file="breakdown_of_pages_using_async_defer.sql") }}</figcaption>
</figure>

> There was an antipattern observed in the last year's results that some websites use both `async` and `defer` attribute resulting in fallback to `async` if the browser supports it.

Reviewer 2: ```suggestion
There was an anti-pattern observed in the last year's results that some websites use both `async` and `defer` attribute resulting in fallback to `async` if the browser supports it.
```

> The usage was so frequent that [11.4%](https://almanac.httparchive.org/en/2020/javascript#how-do-we-load-our-javascript) of requests made on the mobile pages were seen with at least one script with `async` and `defer` attributes used together.  The [root causes](https://twitter.com/rick_viscomi/status/1331735748060524551?s=20) were found and an action item was also taken down to [remove such usage going forward](https://twitter.com/Kraft/status/1336772912414601224?s=20). It’d be very interesting this year to look at where this stands.

Reviewer 2: ```suggestion
The usage was so frequent that [11.4%](https://almanac.httparchive.org/en/2020/javascript#how-do-we-load-our-javascript) of requests made on the mobile pages were seen with at least one script with `async` and `defer` attributes used together. The [root causes](https://twitter.com/rick_viscomi/status/1331735748060524551?s=20) were found and an action item was also taken down to [remove such usage going forward](https://twitter.com/Kraft/status/1336772912414601224?s=20). It’d be very interesting this year to look at where this stands.
```

<script type="text/javascript">
   (function() {
      var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
      g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'piwik.js'; s.parentNode.insertBefore(g,s);
   })();
</script>
<!-- End Piwik Code -->
```

According to [Wikipedia](https://en.wikipedia.org/wiki/Matomo_(software)), Piwik (now known as Matomo):
<blockquote>is a free and open source web analytics application developed by a team of international developers, that runs on a PHP/MySQL webserver. It tracks online visits to one or more websites and displays reports on these visits for analysis. As of June 2018, Matomo was used by over 1,455,000 websites, or 1.3% of all websites with known traffic analysis tools... </blockquote>

This strongly suggests that marketing and analytics providers are a common source of these async/defer scripts being dynamically injected into the page.

{{ figure_markup(
  content="2.6%",
  caption="Percent of requests using the `async` and `defer` attribute together.",
  classes="big-number",
  sheets_gid="655357075",
  sql_file="breakdown_of_scripts_using_async_defer_module_nomodule.sql"
  )
}}

> As it turns out, 2.6% of requests use this antipattern with both `async` and `defer` on the same script element.

Reviewer 2: ```suggestion
As it turns out, 2.6% of requests use this anti-pattern with both `async` and `defer` on the same script element.
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=395424439&format=interactive",
  sheets_gid="1354767842",
  sql_file="requests_by_3p.sql"
  )
}}

A median of mobile pages request 10 third-party resources whereas 9 first-party requests. let's look at how the results vary for the desktop client.

{{ figure_markup(
  image="js-requests-desktop-host.png",
  caption="Distribution of the number of JavaScript requests per desktop page by host.",
  description="Bar chart showing distribution of JavaScript requests per desktop page by host. A median mobile page makes 10 first-party requests and 11 third-party ones. The 10th, 25th, 50th, 75th, and 90th percentiles for requests made on desktop for first-party are: 2, 4, 10, 19, 33.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1584404981&format=interactive",
  sheets_gid="1354767842",
  sql_file="requests_by_3p.sql"
  )
}}

> A median of desktop pages request 11 third-party resources whereas 10 first-party requests. This difference increases as we move up to the 90th percentile as 33 requests on mobile pages are first party but the number goes up to 34 for third-party requests for the mobile pages.  Clearly, the number of third-party resources requested is always one step ahead of the first-party ones. Irrespective of the performance and reliability [risks that requesting third-party resources brings](https://css-tricks.com/potential-dangers-of-third-party-javascript/), the usage seems to favour third-party scripts, which could be due to the [useful interactivity features](https://developers.google.com/web/fundamentals/performance/critical-rendering-path/adding-interactivity-with-javascript) that it gives to the web. 

Reviewer 2: ```suggestion
A median of desktop pages request 11 third-party resources whereas 10 first-party requests. This difference increases as we move up to the 90th percentile as 33 requests on mobile pages are first party but the number goes up to 34 for third-party requests for the mobile pages. Clearly, the number of third-party resources requested is always one step ahead of the first-party ones. Irrespective of the performance and reliability [risks that requesting third-party resources brings](https://css-tricks.com/potential-dangers-of-third-party-javascript/), the usage seems to favor third-party scripts, which could be due to the [useful interactivity features](https://developers.google.com/web/fundamentals/performance/critical-rendering-path/adding-interactivity-with-javascript) that it gives to the web. 
```

Reviewer 2: `favour` vs `favor` feel free to ignore.

50</p>

   </td>
   <td><p style="text-align: right">
4</p>

   </td>
   <td><p style="text-align: right">
4</p>

   </td>
  </tr>
  <tr>
   <td><p style="text-align: right">
75</p>

   </td>
   <td><p style="text-align: right">
8</p>

   </td>
   <td><p style="text-align: right">
8</p>

   </td>
  </tr>
  <tr>
   <td><p style="text-align: right">
90</p>

   </td>
   <td><p style="text-align: right">
15</p>

   </td>
   <td><p style="text-align: right">
16</p>

   </td>
  </tr>
  <tr>
   <td><p style="text-align: right">
100</p>

   </td>
   <td><p style="text-align: right">
623</p>

   </td>
   <td><p style="text-align: right">
867</p>

   </td>
  </tr>
</table>

<figcaption>{{ figure_link(caption="Distribution of the number of asynchronous requests made per page.", sheets_gid="183546956", sql_file="ajax_request_per_page.sql") }}</figcaption>
</figure>

The most asynchronous requests made by a desktop page is 623, which is eclipsed by the biggest mobile page, which makes 867 asynchronous requests!

An alternative to the asynchronous AJAX requests are the synchronous. Rather than passing a request to a callback, they block the main thread until the request completes.

> However, this practice is [discouraged](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Synchronous_and_Asynchronous_Requests#:~:text=Note%3A%20Starting%20with%20Gecko%2030.0%20(Firefox%2030.0%20/%20Thunderbird%2030.0%20/%20SeaMonkey%202.27)%2C%20Blink%2039.0%2C%20and%20Edge%2013%2C%20synchronous%20requests%20on%20the%20main%20thread%20have%20been%20deprecated%20due%20to%20their%20negative%20impact%20on%20the%20user%20experience) due to the potential for poor performance and user experiences, and many browsers already warn about the usage.  It would be intriguing to see how many pages still use synchronous AJAX requests.

Reviewer 2: ```suggestion
However, this practice is [discouraged](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Synchronous_and_Asynchronous_Requests#:~:text=Note%3A%20Starting%20with%20Gecko%2030.0%20(Firefox%2030.0%20/%20Thunderbird%2030.0%20/%20SeaMonkey%202.27)%2C%20Blink%2039.0%2C%20and%20Edge%2013%2C%20synchronous%20requests%20on%20the%20main%20thread%20have%20been%20deprecated%20due%20to%20their%20negative%20impact%20on%20the%20user%20experience) due to the potential for poor performance and user experiences, and many browsers already warn about the usage. It would be intriguing to see how many pages still use synchronous AJAX requests.
```

Reviewer 2: There is a little inconsistency in the spacing here. You can decide if you want to use one or two spaces for new sentences.

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=584103777&format=interactive",
  sheets_gid="1851485826",
  sql_file="frameworks_libraries.sql"
  )
}}

jQuery remains on top with 84% of mobile pages containing jQuery, however React usage has jumped from 4% to 8% since last year, which is a significant increase. Also worth noticing is the usage of Isotope (uses jQuery) to 7% leading to RequireJS falling out of the picture with 2% usage.

One might wonder why jQuery is so dominant and hasn't disappeared over time. There are two main reasons for this. First, as [highlighted over the previous years](https://almanac.httparchive.org/en/2019/javascript#:~:text=There%20are%20a,client-side%20libraries.), is that most of the [WordPress](https://wordpress.org/) sites use jQuery, which contributes to a major participation of jQuery. Second, even if the wide usage of jQuery is ignored for a moment, several of the other top-used JavaScript libraries still rely on jQuery in some way under the hood.

**UI Frameworks**

> The adoption of JavaScript frameworks doesn’t see a substantial change compared to the previous years.  With the way the adoption is measured, there is a [detection limitation](https://github.com/AliasIO/wappalyzer/issues/2450) which doesn’t let Wappalyzer capture the percentage precisely.

Reviewer 2: ```suggestion
The adoption of JavaScript frameworks doesn’t see a substantial change compared to the previous years. With the way the adoption is measured, there is a [detection limitation](https://github.com/AliasIO/wappalyzer/issues/2450) which doesn’t let Wappalyzer capture the percentage precisely. Additionally, the detection is only run on home pages which doesn't capture accurately the overall framework adoption.
```

  </tr>
</table>

<figcaption>{{ figure_link(caption="Javascript frameworks usage with JavaScript libraries.", sheets_gid="1934429143", sql_file="frameworks_libraries_combos.sql") }}</figcaption>
</figure>

As can be observed, jQuery tops in usage with libraries and frameworks, which could also be a result of using a third-party library that includes it. This, as a result, leads to more processing time for the code that the framework or library uses with the other included libraries.

22% of known versions of jQuery are found to be version 3.5.1. This is a big jump compared to last year’s (1.12.4), which could be attributed to Wordpress which constitutes most of the participation of jQuery. 

**Web components and shadow DOM**

With the web becoming componentized, a developer building a single page application may think about a user view as a set of components. This is not only for the sake of developers building dedicated components for each feature, but also to maximize component reusability. It could be in the same app on a different view or in a completely different application. Such use cases lead to the usage of custom elements and web components in web applications. 

> It would be justified to say that with many JavaScript frameworks gaining popularity, the idea of reusability and building dedicated feature-based components has been adopted more widely.  This feeds our curiosity to look into the adoption of custom elements, shadow DOM, template elements.

Reviewer 2: ```suggestion
It would be justified to say that with many JavaScript frameworks gaining popularity, the idea of reusability and building dedicated feature-based components has been adopted more widely. This feeds our curiosity to look into the adoption of custom elements, shadow DOM, template elements.
```

{{ figure_markup(
  content="<0.1%",
  caption="Percent of pages using templates.",
  classes="big-number",
  sheets_gid="1991863136",
  sql_file="web_components_specs.sql"
  )
}}

Less than 0.1% of web pages have adopted the use of templates. Although templates are well [supported](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_templates_and_slots#:~:text=Templates%20are%20well%2Dsupported%20in%20browsers%3B%20the%20Shadow%20DOM%20API%20is%20supported%20by%20default%20in%20Firefox%20(version%2063%20onwards)%2C%20Chrome%2C%20Opera%2C%20Safari%2C%20and%C2%A0Edge) in browsers, there is still a very low percentage of pages using them.

### Preload and Prefetch

At the render of a page, the browser downloads the given resources and to prioritise the download of some resources over others the browser uses [resource hints](https://almanac.httparchive.org/en/2021/resource-hints). The resource hints, preload and prefetch, are basically to tell the browser to either download some resource right away or to download it whenever it can to be able to present the resource when required.

> The **_preload_** hint tells the browser to download the resource with a higher priority as it will be required on the current page.  \

Reviewer 2: ```suggestion
The **_preload_** hint tells the browser to download the resource with a higher priority as it will be required on the current page.
```

> The **_prefetch_** hint however tells the browser that the resource could be required after some time (useful for future navigation) and it’d better to fetch it when the browser has the capacity to do so and make it available as soon as it is required.

Reviewer 2: ```suggestion
The **_prefetch_** hint, however, tells the browser that the resource could be required after some time (useful for future navigation) and it’d better to fetch it when the browser has the capacity to do so and make it available as soon as it is required.
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1306457429&format=interactive",
  sheets_gid="1182320606",
  sql_file="compression_method.sql"
  )
}}

It is noted that 56% of JS requests apply the gzip compression method, whereas 31% of requests from mobile devices have the brotli compression applied. 

{{ figure_markup(
  content="30%",
  caption="Desktop requests apply brotli compression.",
  classes="big-number",
  sheets_gid="1182320606",
  sql_file="compression_method.sql"
  )
}}
  

An interesting observation here is the trend of the usage of these compression methods as [compared to the previous year](https://almanac.httparchive.org/en/2019/javascript#fig-10). The usage of gzip has gone down by almost 10% and brotli has increased by 11%. The trend definitely indicates the focus on smaller size files with more levels of compression that Brotli provides as compared to gzip.

> With such comparable data, let us actually check if the percentage difference is somehow impacted by the resources being first party or third party.

Reviewer 2: ```suggestion
With such comparable data, let us actually check if the percentage difference is somehow impacted by the resources being first party or third-party.
```

  sheets_gid="1539653841",
  sql_file="lighthouse_unminified_js.sql"
  )
}}

It is noted that 67% of mobile pages have an "unminified JS" score between 0.9 and 1.0, meaning that they have a few unminified JS resources.

 This is an improvement of 10% as compared to the results from the last year (2020).

To dive deeper into understanding how many bytes per page are unminified.

{{ figure_markup(
  image="unminified-js-bytes.png",
  caption="Percentage distribution of unminified JS bytes per page.",
  description="Bar chart showing the percentage distribution of unminified JS bytes per page. 57% of mobile pages have 0 KB of unminified JS whereas 18% of those have 0-10 kilobytes of unminified JS.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1069289665&format=interactive",
  sheets_gid="1511556407",
  sql_file="lighthouse_unminified_js_bytes.sql"
  )
}}

And it is found that 57% of mobile pages have 0 KB of unminified JS whereas 18% of those have 0-10 kilobytes of unminified JS.

> The first party vs. third party analysis in this case shows that 82% of the average mobile page's unminified JS bytes actually come from first-party scripts.

Reviewer 2: ```suggestion
The first-party vs. third-party analysis in this case shows that 82% of the average mobile page's unminified JS bytes actually come from first-party scripts.
```

 \
One step towards improving the percentage is to look at the [percentage of unused JavaScript](https://web.dev/unused-javascript/) when measuring the lighthouse report for the page and [finding opportunities to get rid of the unnecessary JavaScript.](https://web.dev/remove-unused-code/)

### JS requests per page

One of the contributing factors towards slow rendering of the web page could be the requests made on the page, especially when they are blocking.  It would be interesting to look at the number of JavaScript requests made per page on both desktop and mobile devices.

{{ figure_markup(
  image="js-requests-per-page.png",
  caption="Distribution of JavaScript requests on desktop and mobile pages.",
  description="",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1487796178&format=interactive",
  sheets_gid="159538568",
  sql_file="requests_2021.sql"
  )
}}

The median desktop page loads 21 JavaScript resources(.js and .mjs files), going up to 59 resources at the 90th percentile. 

<figure markdown>
<table>
  <tr>

>    <td><em>percentile</em>

Reviewer: Yes, we can update this in editing though. We'll use `th` per the [figure guide](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Figures-Guide#complex-html-table-example).

>    <td><em>percentile</em>

Reviewer 3: Should I resolve this one then?

>    <td><em>percentile</em>

Reviewer: Yeah I'll resolve it

      g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'piwik.js'; s.parentNode.insertBefore(g,s);
   })();
</script>
<!-- End Piwik Code -->
```

According to [Wikipedia](https://en.wikipedia.org/wiki/Matomo_(software)), Piwik (now known as Matomo):
<blockquote>is a free and open source web analytics application developed by a team of international developers, that runs on a PHP/MySQL webserver. It tracks online visits to one or more websites and displays reports on these visits for analysis. As of June 2018, Matomo was used by over 1,455,000 websites, or 1.3% of all websites with known traffic analysis tools... </blockquote>

This strongly suggests that marketing and analytics providers are a common source of these async/defer scripts being dynamically injected into the page.

{{ figure_markup(
  content="2.6%",
  caption="Percent of requests using the `async` and `defer` attribute together.",
  classes="big-number",
  sheets_gid="655357075",
  sql_file="breakdown_of_scripts_using_async_defer_module_nomodule.sql"
  )
}}

As it turns out, 2.6% of requests use this antipattern with both `async` and `defer` on the same script element.

### first party vs third party

> The median number of JavaScript resources requested on desktop pages, as [observed earlier](#how-much-javascript-do-we-load), is 21. Let us now look into which ones of these are first-party or third-party requests. 

Author: Added text between the charts to explain, not changing the title to keep the charts consistent.

An AJAX request works with different formats of data that can be sent and received. When a resource is requested, there is more significant information that is sent to the server to make the results specific and clear. One such requirement is sending the correct required content types. The request sent from the browser requests a content type that gets back the data in the requested format.

Out of the vast list of content types that can be requested, let's look at the most commonly requested content types. The commonly requested resources are  JavaScript, images, etc.

The most requested content types on mobile and desktop pages are images at 45% of requests. Images are a broad category inclusive of all the content types related to images, for example, `image/x-icon`, `image/svg+xml`, `image/webp`, `image/jpeg`, etc.

{{ figure_markup(
  content="45.6%",
  caption="Percent of AJAX requests for images.",
  classes="big-number",
  sheets_gid="538991873",
  sql_file="most_requested_content_type.sql"
  )
}}

<figure markdown>

> <table>

Author: Yes, that's a good point, sorted and changed column name.

{{ figure_markup(
  image="ajax_xhr.png",
  caption="Usage of XMLHttpRequest",
  description="",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1164635195&format=interactive",
  sheets_gid="1012977216",
  sql_file="percentage_usage_of_different_ajax_apis.sql"
  )
}}

The median mobile page makes 2 XHR requests, but at 90th percentile, makes 6 XHR requests. 

{{ figure_markup(
  image="ajax_fetch.png",
  caption="Usage of Fetch",
  description="",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=147712640&format=interactive",
  sheets_gid="1012977216",
  sql_file="percentage_usage_of_different_ajax_apis.sql"
  )
}}

In the case of the usage of the Fetch API, a median mobile page makes 2 requests per page and in the long tail, reaches 3 requests. This is reaching towards the standard XHR way of making requests and in turn, with a much cleaner approach and less boilerplate code, the overall load time will improve too.

> One major reason for the lower usage of Beacon could be that it is typically used for sending analytics data, that too in the cases where one wants to ensure the request is sent even if the page might unload soon. This is, however, not guaranteed when using XHR. A good experiment for the future would be to see if some statistics could be collected around any pages using XHR for analytics data, session data etc.

Author: >I think it's the "that too" that confuses me.

made some changes to express it better, have a look :)

{{ figure_markup(
  image="ajax_beacon.png",
  caption="Usage of Beacon",
  description="",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1483701430&format=interactive",
  sheets_gid="1012977216",
  sql_file="percentage_usage_of_different_ajax_apis.sql"
  )
}}

It would be interesting to also compare the adoption of XHR and Fetch over time.

<figure markdown>

> <table>

Author: Did a bit of a workaround to keep both for detail. Have a look.

  sheets_gid="821396474",
  sql_file="compression_method_by_3p.sql"
  )
}}

59% of third-party scripts are gzipped and 30% are brotli compressed. Looking at first-party scripts, these are 52% with gzip compression but only 32% with brotli. There are still 11% of third-party scripts that do not have any compression method defined for the resources.

{{ figure_markup(
  image="uncompressed-first-third-party.png",
  caption="Uncompressed resources for first party vs third party.",
  description="",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1043138204&format=interactive",
  sheets_gid="1942216454",
  sql_file="compression_none_by_bytes.sql"
  )
}}

The way preloading all our requests negates the impact of the attribute, a similar problem occurs when we try to optimise and compress some resource that doesn’t need compression and is already small. It is observed that 90% of uncompressed third party JS resources are &lt; 5 KB, and uncompressed first-party resources less than 10kb are only 8%. 

### Minification

> When working towards reducing the script parsing time, we tend to focus on all the opportunities to make our code smaller and relevant. One such idea is to minify the files and bring down the bytes usage.

Reviewer 3: relevant doesn't seem the right adjective here. How about **more efficient**?

  caption="Percentage distribution of unminified JS audit scores.",
  description="",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1572896641&format=interactive",
  sheets_gid="1539653841",
  sql_file="lighthouse_unminified_js.sql"
  )
}}

It is noted that 67% of mobile pages have an "unminified JS" score between 0.9 and 1.0, meaning that they have a few unminified JS resources.

 This is an improvement of 10% as compared to the results from the last year (2020).

To dive deeper into understanding how many bytes per page are unminified.

{{ figure_markup(
  image="unminified-js-bytes.png",
  caption="Percentage distribution of unminified JS bytes per page.",
  description="",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1069289665&format=interactive",
  sheets_gid="1511556407",
  sql_file="lighthouse_unminified_js_bytes.sql"
  )
}}

And it is found that 57% of mobile pages have 0 KB of unminified JS whereas 18% of those have 0-10 kilobytes of unminified JS.

> The first party vs. third party analysis in this case shows that 82% of the average mobile page's unminified JS bytes actually come from first-party scripts.

Reviewer 3: That's not what the image below shows. It shows 82% of unminified bytes come from third parties.

Author: Thanks for catching this. Fixed!

{{ figure_markup(
  image="average-unminified-js-bytes.png",
  caption="Average distribution of unminified JS Bytes.",
  description="",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1119550643&format=interactive",
  sheets_gid="127709080",
  sql_file="lighthouse_unminified_js_by_3p.sql"
  )
}}

### Source maps

Source maps are files sent along with the JavaScript resource files to let the browser map the minified or compressed file to their original version of that file. This is how source maps help us in debugging in the production environment. 

{{ figure_markup(
  content="0.09%",
  caption="Mobile pages using the sourcemap header.",
  classes="big-number",
  sheets_gid="1186092564",
  sql_file="sourcemap_header.sql"
  )
}}

> It is found that ~98% of mobile pages do not use the SourceMap response header on script resources whereas only 0.09% use it on resources. One reason for this extremely small percentage could be that not many sites choose to put their source code in production through the source map.

Reviewer 3: "...put their **original** source code..." instead of "...put their source code..."

<figure>
  <table>
    <thead>
      <tr>
        <th>Client</th>
        <th>`async`</th>
        <th>`defer`</th>
        <th>Neither</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>Desktop</td>
        <td class="numeric">89.3%</td>
        <td class="numeric">48.1%</td>
        <td class="numeric">10.3%</td>
      </tr>
      <tr>
        <td>Mobile</td>
        <td class="numeric">89.1%</td>
        <td class="numeric">47.8%</td>
        <td class="numeric">10.4%</td>
      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(caption="Percent of pages using `async`, `defer`, or neither.", sheets_gid="2038121216", sql_file="breakdown_of_pages_using_async_defer.sql") }}</figcaption>
</figure>

> There was an anti-pattern observed in the last year's results that some websites use both `async` and `defer` attribute resulting in fallback to `async` if the browser supports it. {# TODO Why is it an anti-pattern? It is because async takes precendence - we should explain this. Also I presume it is still the case? #}

Reviewer 4: @NishuGoel

A typical <a hreflang="en" href="https://www.tntsupercenter.com/">example website</a>, loads the below script:

```js
<!-- Piwik -->
<script type="text/javascript">
   (function() {
      var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
      g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'piwik.js'; s.parentNode.insertBefore(g,s);
   })();
</script>
<!-- End Piwik Code -->
```

According to [Wikipedia](https://en.wikipedia.org/wiki/Matomo_(software)), Piwik (now known as Matomo):
<blockquote>is a free and open source web analytics application developed by a team of international developers, that runs on a PHP/MySQL webserver. It tracks online visits to one or more websites and displays reports on these visits for analysis. As of June 2018, Matomo was used by over 1,455,000 websites, or 1.3% of all websites with known traffic analysis tools...</blockquote>

This strongly suggests that marketing and analytics providers are a common source of these async/defer scripts being dynamically injected into the page.

> {# TODO Why does this strongly suggest this? Is it the fact that they are obviously dynamically inserted? Is it only marketing and analytics providers that prodominately using dynamically inserted script tags like this? #}

Reviewer 4: @NishuGoel

Author: tweaked the sentence a bit to make it sound simpler

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1180601651&format=interactive",
  sheets_gid="1107832831",
  sql_file="resource-hints-prefetch-preload-percentage.sql"
  )
}}

A median desktop page loads 1 JavaScript resource with the preload hint whereas 2 JavaScript resources with the prefetch hint.

<figure>
  <table>
    <thead>
      <tr>
        <th>Client</th>
        <th>2020</th>
        <th>2021</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>Desktop</td>
        <td class="numeric">5</td>
        <td class="numeric">4</td>
      </tr>
      <tr>
        <td>Mobile</td>
        <td class="numeric">7</td>
        <td class="numeric">4</td>
      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(caption="Year over year comparison of the usage of preload hints per page at the 90th percentile.", sheets_gid="1107832831", sql_file="resource-hints-prefetch-preload-percentage.sql") }}</figcaption>
</figure>

> {# TODO why the 90th percentile? Would normally advise to look at the median #}

Reviewer 4: @NishuGoel

<figure>
  <table>
    <thead>
      <tr>
        <th>Client</th>
        <th>2020</th>
        <th>2021</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>Desktop</td>
        <td class="numeric">14</td>
        <td class="numeric">13</td>
      </tr>
      <tr>
        <td>Mobile</td>
        <td class="numeric">12</td>
        <td class="numeric">11</td>
      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(caption="Year over year comparison of the usage of prefetch hints per page at the 90th percentile.", sheets_gid="1107832831", sql_file="resource-hints-prefetch-preload-percentage.sql") }}</figcaption>
</figure>

> {# TODO same questions #}

Reviewer 4: @NishuGoel

When working towards reducing the script parsing time, we tend to focus on all the opportunities to make our code smaller and more efficient. One such idea is to minify the files and bring down the bytes usage.

The Lighthouse report also <a hreflang="en" href="https://web.dev/unminified-javascript/">highlights the unminified JS being used</a> and lists these unminified files and the opportunities to save.

{{ figure_markup(
  image="unminified-js-audit-scores.png",
  caption="Percentage distribution of unminified JS audit scores.",
  description='Bar chart showing the percentage distribution of unminified JS audit scores. 67% of mobile pages have an "unminified JS" score between 0.9 and 1.0',
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1572896641&format=interactive",
  sheets_gid="1539653841",
  sql_file="lighthouse_unminified_js.sql"
  )
}}

It is noted that 67.1% of mobile pages have an "unminified JS" score between 0.9 and 1.0, meaning that they have a few unminified JS resources.

 This is an improvement of 10% as compared to the results from the last year (2020).

>  {# TODO Isn't this a regression of 10%? #}

Reviewer 4: @NishuGoel

Author: fixed!

To dive deeper into understanding how many bytes per page are unminified.

{{ figure_markup(
  image="unminified-js-bytes.png",
  caption="Percentage distribution of unminified JS bytes per page.",
  description="Bar chart showing the percentage distribution of unminified JS bytes per page. 57% of mobile pages have 0 KB of unminified JS whereas 18% of those have 0-10 kilobytes of unminified JS.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1069289665&format=interactive",
  sheets_gid="1511556407",
  sql_file="lighthouse_unminified_js_bytes.sql"
  )
}}

And it is found that 57.4% of mobile pages have 0 KB of unminified JS whereas 17.9% of those have 0-10 kilobytes of unminified JS.

> {# TODO I don't really understand what this graph is showing me. Especially compared to last graph. These sound small so is this good? Or bad? #}

Reviewer 4: @NishuGoel

Author: explained what's happening, please review

  classes="big-number",
  sheets_gid="1186092564",
  sql_file="sourcemap_header.sql"
  )
}}

It is found that only 0.1% of mobile pages do not use the SourceMap response header on script resources. One reason for this extremely small percentage could be that not many sites choose to put their original source code in production through the source map.

An analysis of how many sites actually send the source map header on their first-party or third-party scripts shows:

{{ figure_markup(
  image="sourcemap-first-third-party.png",
  caption="Usage of sourcemap headers in first party vs third party.",
  description="Bar chart showing the usage of sourcemap headers in first party vs third party. 98% of the JS requests that include a SourceMap header on mobile are for first-party scripts and only 2% for third party",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1762186682&format=interactive",
  sheets_gid="2057978707",
  sql_file="sourcemap_header_by_3p.sql"
  )
}}

98% of the JS requests that include a SourceMap header on mobile are for first-party scripts.

> {# TODO That seems surprising to me. I would have though it was more common to be from third-parties since the likes of query and Bootstrap include source map files. Any thoughts on why this is the reverse? Or am I just wrong about this expectation? #}

Reviewer 4: @NishuGoel

Author: added to the comment, please review

Reviewer: As I'm editing the chapter I found the TODO and this discussion thread.

98% comes from the [pivot table](https://docs.google.com/spreadsheets/d/1zU9rHpI3nC6jTz3xgN6w13afW7x34xAKBh2IPH-lVxk/edit?disco=AAAAQ5pjLK8) that I've added to the sheet. Of the requests that have the header set, 230,623 are first-party and 4,745 are third-party, or 98% first-party.

Here's the [query](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2021/javascript/sourcemap_header_by_3p.sql) for reference.

I don't think the way it was rewritten is correct:

> An analysis of how many sites actually send the source map header on their first-party or third-party scripts shows that 0.1% of the JavaScript requests that include a SourceMap header on mobile are for first-party scripts.

0.1 is not the % of JS requests with the header, but % of all JS requests.

Author: Sure @rviscomi 
We discussed it later. 
Could you revert it please to 98%?

        <td>Knockout</td>
        <td class="numeric">0.2%</td>
      </tr>
      <tr>
        <td>Lo-Dash</td>
        <td class="numeric">3.1%</td>
      </tr>
      <tr>
        <td>Moment.js</td>
        <td class="numeric">2.2%</td>
      </tr>
      <tr>
        <td>Mustache</td>
        <td class="numeric">0.7%</td>
      </tr>
      <tr>
        <td>Underscore</td>
        <td class="numeric">6.4%</td>
      </tr>
      <tr>
        <td>Vue</td>
        <td class="numeric">0.2%</td>
      </tr>
      <tr>
        <td>Vue (Fast path)</td>
        <td class="numeric">0.2%</td>
      </tr>
      <tr>
        <td>React</td>
        <td class="numeric">0%</td>
      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(caption="List of security vulnerable libraries with their vulnerability percentage.", sheets_gid="551476210", sql_file="lighthouse_vulnerable_libraries.sql") }}</figcaption>
</figure>

The results show that jQuery is the dominant factor in the decrease of vulnerability with at least 23.3% of improvement here and some improvement in jQueryUI too. However, there is little to no difference in other libraries' security vulnerabilities, except for a few.

> {# TODO which few? Or should we just remove that "except for a few" part? Or should we update the table to include the improvement? #}

Reviewer 4: @NishuGoel
